dali/pipeline/operators/arithmetic/expression_tree.cc

```diff
@@ -0,0 +1,170 @@
+  EnforceNonEnd(expr, pos, "function name or input description starting with \"&\" or \"$\"");
+  DALI_ENFORCE(
+      std::isalpha(expr[pos]),
+      make_string_delim(
+          "", "Unrecognized token for expression description: ", ReportCharacter(expr[pos]),
+          " at position [", pos, "] in: \"", expr,
+          "\". Expected function name starting with alphabetic character or input description "
+          "starting with \"&\" or \"$\"."));
+  int end_pos = pos + 1;
+  while (end_pos < static_cast<int>(expr.length()) && std::isalnum(expr[end_pos])) {
+    end_pos++;
+  }
+  return std::make_pair(std::string(&expr[pos], &expr[end_pos]), end_pos);
+}
+
+ParseResult ParseCall(const std::string &expr, int pos) {
+  std::string op;
+  std::tie(op, pos) = ParseName(expr, pos);
+  auto node = std::make_unique<ExprFunc>(op);
+  pos = SkipSpaces(expr, pos, "opening parenthesis \"(\" for subexpression");
+  pos = ExpectChar(expr, pos, '(');
+  pos = SkipSpaces(expr, pos, "contents of subexpression or closing parenthesis \")\"");
+  while (expr[pos] != ')') {
+    std::unique_ptr<ExprNode> subexpr;
+    std::tie(subexpr, pos) = ParseExpr(expr, pos);
+    node->AddSubexpression(std::move(subexpr));
+    pos = SkipSpaces(expr, pos, "contents of next subexpression or closing parenthesis \")\"");
+  }
+  pos = ExpectChar(expr, pos, ')');
+  return std::make_tuple(std::move(node), pos);
+}
+
+ParseResult ParseInput(const std::string &expr, int pos) {
+  int mapped_input;
+  std::tie(mapped_input, pos) = ParseInt(expr, pos);
+  auto node = std::make_unique<ExprTensor>(mapped_input);
+  return std::make_tuple(std::move(node), pos);
+}
+
+ParseResult ParseConstant(const std::string &expr, int pos) {
+  // DALI_FAIL("Constant description in expressions is not supported.");
+  int mapped_input;
+  std::tie(mapped_input, pos) = ParseInt(expr, pos);
+  pos = ExpectChar(expr, pos, ':');
+  // TODO(klecki): Parse the type encoding
+  DALIDataType type = DALIDataType::DALI_INT32;
+  while (std::isalnum(expr[pos])) {
+    pos++;
+  }
+  auto node = std::make_unique<ExprConstant>(mapped_input, type);
+  return std::make_tuple(std::move(node), pos);
+}
+
+ParseResult ParseExpr(const std::string &expr, int pos) {
+  EnforceNonEnd(expr, pos, "function name or input description starting with \"&\" or \"$\"");
+  if (expr[pos] == '&') {
+    return ParseInput(expr, pos + 1);
+  } else if (expr[pos] == '$') {
+    return ParseConstant(expr, pos + 1);
+  } else {
+    return ParseCall(expr, pos);
+  }
+}
+
+}  // namespace
+
+/**
+ * @brief Parse `expression_desc` provided to ArithmeticGenericOp.
+ *
+ * @param expr Expression string as specified in Schema for ArithmeticGenericOp
+ * @return parsed epxression tree, to be filled with tensor types and shapes
+ */
+std::unique_ptr<ExprNode> ParseExpressionString(const std::string &expr) {
+  return std::move(std::get<0>(ParseExpr(expr, 0)));
+}
+
+}  // namespace dali
```

Comment:
Unit-tests for `ParseExpressionString` is a must. I get that you don't like TDD and stuff, in that case please provide good docs for rest of the functions (so whoever looks at it doesn't need to spend too much time in understanding what's going on there) and put some `asserts`, so that in case of doubt it can be run in debug and checked, that everything is OK

Reply:
Moved from arithmetic_test.